src/integer/z/fmpz_helpers.rs

```diff
@@ -74,6 +77,196 @@ pub(crate) fn distance(value_1: &fmpz, value_2: &fmpz) -> Z {
+                let mut ret_value = fmpz(0);
+                fmpz_init_set_si(&mut ret_value, *self as i64);
+                ret_value
+            }
+        }
+    )*
+    };
+}
+
+AsInteger_singed!(i8 u8 i16 u16 i32 u32 i64);
+
+unsafe impl AsInteger for Z {
+    /// Documentation at [`AsInteger::into_fmpz`]
+    unsafe fn into_fmpz(mut self) -> fmpz {
+        let mut out = fmpz(0);
+        fmpz_swap(&mut out, &mut self.value);
+        out
+    }
+
+    /// Documentation at [`AsInteger::get_fmpz_ref`]
+    fn get_fmpz_ref(&self) -> Option<&fmpz> {
+        Some(&self.value)
+    }
+}
+
+unsafe impl AsInteger for &Z {
+    /// Documentation at [`AsInteger::into_fmpz`]
+    unsafe fn into_fmpz(self) -> fmpz {
+        let mut value = fmpz(0);
+        fmpz_set(&mut value, &self.value);
+        value
+    }
+
+    /// Documentation at [`AsInteger::get_fmpz_ref`]
+    fn get_fmpz_ref(&self) -> Option<&fmpz> {
+        Some(&self.value)
+    }
+}
+
+#[cfg(test)]
+mod test_as_integer_rust_ints {
+    use crate::{integer::Z, traits::AsInteger};
+
+    /// Assert that rust integers don't have an internal fmpz
+    #[test]
+    fn get_fmpz_ref_none() {
+        assert!(10u8.get_fmpz_ref().is_none());
+        assert!(10u16.get_fmpz_ref().is_none());
+        assert!(10u32.get_fmpz_ref().is_none());
+        assert!(10u64.get_fmpz_ref().is_none());
+        assert!(10i8.get_fmpz_ref().is_none());
+        assert!(10i16.get_fmpz_ref().is_none());
+        assert!(10i32.get_fmpz_ref().is_none());
+        assert!(10i64.get_fmpz_ref().is_none());
+    }
+
+    /// Assert that into_fmpz works correctly for small values
+    #[test]
+    fn into_fmpz_small() {
+        let z = Z::from(10);
+
+        unsafe {
+            assert_eq!(z, Z::from_fmpz(10u8.into_fmpz()));
+            assert_eq!(z, Z::from_fmpz(10u16.into_fmpz()));
+            assert_eq!(z, Z::from_fmpz(10u32.into_fmpz()));
+            assert_eq!(z, Z::from_fmpz(10u64.into_fmpz()));
+            assert_eq!(z, Z::from_fmpz(10i8.into_fmpz()));
+            assert_eq!(z, Z::from_fmpz(10i16.into_fmpz()));
+            assert_eq!(z, Z::from_fmpz(10i32.into_fmpz()));
+            assert_eq!(z, Z::from_fmpz(10i64.into_fmpz()));
+        }
+    }
+}
+
+#[cfg(test)]
+mod test_as_integer_z {
+    use flint_sys::fmpz::{fmpz_clear, fmpz_set_ui};
+
+    use super::*;
+
+    // Assert that the new fmpz is not related to the old one
+    #[test]
+    fn small_into_fmpz() {
+        let mut value = unsafe {
+            let z = Z::from(42);
+            (&z).into_fmpz()
+        }; // z is dropped here
+
+        // create a new `Z` to potentially overwrite the memory
+        let _ = Z::from(12);
+
+        let copy = Z::from_fmpz(value);
+        assert_eq!(copy, Z::from(42));
+    }
+
+    /// Assert that the new [`fmpz`] is not effected by the original for large numbers.
+    /// This can not be tested for an owned [`Z`], since that would violate the
+    /// ownership constrain -> not compiling.
```

Comment:
I don't understand why this shouldn't work for owned types as the `z` is dropped afterward anyways.
```suggestion
    /// This can not be tested for an owned [`Z`], since that would violate the
    /// ownership constraint -> not compiling.
```

Reply:
You are right. That was still from an old implementation where I compared to the original.
I also added tests for references now.